editor/src/messages/portfolio/document/node_graph/node_graph_message_handler.rs

```diff
@@ -1457,100 +1475,226 @@ impl NodeGraphMessageHandler {
+		navigation_handler: &NavigationMessageHandler,
+	) {
+		let Some(node_graph_ptz) = network_interface.node_graph_ptz(breadcrumb_network_path) else {
+			log::error!("Could not get node graph PTZ");
+			return;
+		};
+
+		let widgets = vec![
+			NumberInput::new(Some(graph_fade_artwork_percentage))
+				.percentage()
+				.display_decimal_places(0)
+				.label("Fade Artwork")
+				.tooltip("Opacity of the graph background that covers the artwork")
+				.on_update(move |number_input: &NumberInput| {
+					DocumentMessage::SetGraphFadeArtwork {
+						percentage: number_input.value.unwrap_or(graph_fade_artwork_percentage),
+					}
+					.into()
+				})
+				.widget_holder(),
+			Separator::new(SeparatorType::Unrelated).widget_holder(),
+			IconButton::new("ZoomIn", 24)
+				.tooltip("Zoom In")
+				.tooltip_shortcut(action_keys!(NavigationMessageDiscriminant::CanvasZoomIncrease))
+				.on_update(|_| NavigationMessage::CanvasZoomIncrease { center_on_mouse: false }.into())
+				.widget_holder(),
+			IconButton::new("ZoomOut", 24)
+				.tooltip("Zoom Out")
+				.tooltip_shortcut(action_keys!(NavigationMessageDiscriminant::CanvasZoomDecrease))
+				.on_update(|_| NavigationMessage::CanvasZoomDecrease { center_on_mouse: false }.into())
+				.widget_holder(),
+			IconButton::new("ZoomReset", 24)
+				.tooltip("Reset Tilt and Zoom to 100%")
+				.tooltip_shortcut(action_keys!(NavigationMessageDiscriminant::CanvasTiltResetAndZoomTo100Percent))
+				.on_update(|_| NavigationMessage::CanvasTiltResetAndZoomTo100Percent.into())
+				.disabled((node_graph_ptz.zoom() - 1.).abs() < 1e-4)
+				.widget_holder(),
+			PopoverButton::new()
+				.popover_layout(vec![
+					LayoutGroup::Row {
+						widgets: vec![TextLabel::new("Node Graph Navigation").bold(true).widget_holder()],
+					},
+					LayoutGroup::Row {
+						widgets: vec![TextLabel::new(
+							"
+								Interactive controls in this\n\
+								menu are coming soon.\n\
+								\n\
+								Pan:\n\
+								• Middle Click Drag\n\
+								\n\
+								Zoom:\n\
+								• Shift + Middle Click Drag\n\
+								• Ctrl + Scroll Wheel Roll
+							"
+							.trim(),
+						)
+						.multiline(true)
+						.widget_holder()],
+					},
+				])
+				.widget_holder(),
+			Separator::new(SeparatorType::Related).widget_holder(),
+			NumberInput::new(Some(navigation_handler.snapped_zoom(node_graph_ptz.zoom()) * 100.))
+				.unit("%")
+				.min(0.000001)
+				.max(1000000.)
+				.tooltip("Node graph zoom")
+				.on_update(|number_input: &NumberInput| {
+					NavigationMessage::CanvasZoomSet {
+						zoom_factor: number_input.value.unwrap() / 100.,
+					}
+					.into()
+				})
+				.increment_behavior(NumberInputIncrementBehavior::Callback)
+				.increment_callback_decrease(|_| NavigationMessage::CanvasZoomDecrease { center_on_mouse: false }.into())
+				.increment_callback_increase(|_| NavigationMessage::CanvasZoomIncrease { center_on_mouse: false }.into())
+				.widget_holder(),
```

Comment:
Seeing as this is the same as the document navigation overlays (but with the text changed to node graph and using the node graph PTZ struct) and I don't really expect these to diverge (changes to the PTZ system would impact both the document and node graph), would it be preferable to use a shared function?

Reply:
Good point, I'll see if that's workable.